File: lib/swimRecords.js
import { formatDateShort } from './swimFormatters.js';
import { getStatsSessions } from './swimAnalysis.js';

const bestSession = (sessions, pickValue, better) => {
  let best = null;
  sessions.forEach((session) => {
    const value = pickValue(session);
    if (value == null || !Number.isFinite(value)) return;
    if (!best || better(value, best.value)) {
      best = { value, sessionId: session.id, date: session.date };
    }
  });
  return best;
};

export const getPersonalRecords = (sessions) => {
  const statsSessions = getStatsSessions(sessions);
  if (!statsSessions.length) return null;

  return {
    longestDistance: bestSession(
      statsSessions,
      (s) => s.metrics?.distanceM,
      (a, b) => a > b
    ),
    fastestPace: bestSession(
      statsSessions,
      (s) => s.metrics?.paceSecPer100m,
      (a, b) => a < b
    ),
    mostActiveCalories: bestSession(
      statsSessions,
      (s) => s.metrics?.activeKcal,
      (a, b) => a > b
    ),
    mostTotalCalories: bestSession(
      statsSessions,
      (s) => s.metrics?.totalKcal,
      (a, b) => a > b
    ),
    mostLaps: bestSession(
      statsSessions,
      (s) => s.metrics?.laps,
      (a, b) => a > b
    ),
    longestDuration: bestSession(
      statsSessions,
      (s) => s.metrics?.durationSec,
      (a, b) => a > b
    ),
    highestHeartRate: bestSession(
      statsSessions,
      (s) => s.metrics?.avgHeartRate,
      (a, b) => a > b
    ),
  };
};

export const formatRecordDate = (date, locale = 'nl-NL') => {
  if (!date) return '—';
  return formatDateShort(date, locale);
};

File: ios/AapSC/Lib/SwimRecords.swift
import Foundation

enum SwimRecords {
    static func getPersonalRecords(_ sessions: [SwimSession]) -> PersonalRecords? {
        let stats = SwimAnalysis.statsSessions(sessions)
        guard !stats.isEmpty else { return nil }

        return PersonalRecords(
            longestDistance: bestSession(stats, { Double($0.metrics.distanceM ?? 0) }, better: >),
            fastestPace: bestSession(stats, { Double($0.metrics.paceSecPer100m ?? 0) }, better: { $0 < $1 && $0 > 0 }),
            mostActiveCalories: bestSession(stats, { Double($0.metrics.activeKcal ?? 0) }, better: >),
            mostTotalCalories: bestSession(stats, { Double($0.metrics.totalKcal ?? 0) }, better: >),
            mostLaps: bestSession(stats, { Double($0.metrics.laps ?? 0) }, better: >),
            longestDuration: bestSession(stats, { Double($0.metrics.durationSec ?? 0) }, better: >),
            highestHeartRate: bestSession(stats, { Double($0.metrics.avgHeartRate ?? 0) }, better: >)
        )
    }

    private static func bestSession(
        _ sessions: [SwimSession],
        _ pickValue: (SwimSession) -> Double,
        better: (Double, Double) -> Bool
    ) -> PersonalRecord? {
        var best: PersonalRecord?
        for session in sessions {
            let value = pickValue(session)
            guard value.isFinite, value > 0 else { continue }
            if best == nil || better(value, best!.value) {
                best = PersonalRecord(value: value, sessionId: session.id, date: session.date)
            }
        }
        return best
    }
}
